File: mistral/typescript/src/types/embeddings.ts
/**
 * Embedding types.
 */

/**
 * Embedding request.
 */
export interface EmbeddingRequest {
  /** Model ID to use. */
  model: string;
  /** Input text(s) to embed. */
  input: EmbeddingInput;
  /** Encoding format. */
  encoding_format?: EncodingFormat;
}

/**
 * Input for embedding requests.
 */
export type EmbeddingInput = string | string[];

/**
 * Encoding format for embeddings.
 */
export type EncodingFormat = 'float' | 'base64';

/**
 * Embedding response.
 */
export interface EmbeddingResponse {
  /** Response ID. */
  id: string;
  /** Object type. */
  object: string;
  /** Model used. */
  model: string;
  /** Embedding data. */
  data: EmbeddingData[];
  /** Token usage. */
  usage: EmbeddingUsage;
}

/**
 * Embedding data.
 */
export interface EmbeddingData {
  /** Object type. */
  object: string;
  /** Embedding vector. */
  embedding: number[];
  /** Index in the input. */
  index: number;
}

/**
 * Embedding usage information.
 */
export interface EmbeddingUsage {
  /** Number of prompt tokens. */
  prompt_tokens: number;
  /** Total tokens. */
  total_tokens: number;
}

/**
 * Creates an embedding request.
 */
export function createEmbeddingRequest(
  model: string,
  input: EmbeddingInput,
  encodingFormat?: EncodingFormat
): EmbeddingRequest {
  const request: EmbeddingRequest = { model, input };
  if (encodingFormat) {
    request.encoding_format = encodingFormat;
  }
  return request;
}

File: mistral/rust/src/types/embeddings.rs
//! Embedding types.

use serde::{Deserialize, Serialize};

/// Embedding request.
#[derive(Debug, Clone, Serialize)]
pub struct EmbeddingRequest {
    /// Model ID to use.
    pub model: String,
    /// Input text(s) to embed.
    pub input: EmbeddingInput,
    /// Encoding format.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub encoding_format: Option<EncodingFormat>,
}

impl EmbeddingRequest {
    /// Creates a new embedding request.
    pub fn new(model: impl Into<String>, input: impl Into<EmbeddingInput>) -> Self {
        Self {
            model: model.into(),
            input: input.into(),
            encoding_format: None,
        }
    }

    /// Sets the encoding format.
    pub fn with_encoding_format(mut self, format: EncodingFormat) -> Self {
        self.encoding_format = Some(format);
        self
    }
}

/// Input for embedding requests.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum EmbeddingInput {
    /// Single text input.
    Single(String),
    /// Multiple text inputs.
    Multiple(Vec<String>),
}

impl From<String> for EmbeddingInput {
    fn from(s: String) -> Self {
        EmbeddingInput::Single(s)
    }
}

impl From<&str> for EmbeddingInput {
    fn from(s: &str) -> Self {
        EmbeddingInput::Single(s.to_string())
    }
}

impl From<Vec<String>> for EmbeddingInput {
    fn from(v: Vec<String>) -> Self {
        EmbeddingInput::Multiple(v)
    }
}

impl From<Vec<&str>> for EmbeddingInput {
    fn from(v: Vec<&str>) -> Self {
        EmbeddingInput::Multiple(v.into_iter().map(String::from).collect())
    }
}

/// Encoding format for embeddings.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum EncodingFormat {
    /// Float format.
    Float,
    /// Base64 format.
    Base64,
}

/// Embedding response.
#[derive(Debug, Clone, Deserialize)]
pub struct EmbeddingResponse {
    /// Response ID.
    pub id: String,
    /// Object type.
    pub object: String,
    /// Model used.
    pub model: String,
    /// Embedding data.
    pub data: Vec<EmbeddingData>,
    /// Token usage.
    pub usage: EmbeddingUsage,
}

/// Embedding data.
#[derive(Debug, Clone, Deserialize)]
pub struct EmbeddingData {
    /// Object type.
    pub object: String,
    /// Embedding vector.
    pub embedding: Vec<f32>,
    /// Index in the input.
    pub index: u32,
}

/// Embedding usage information.
#[derive(Debug, Clone, Deserialize)]
pub struct EmbeddingUsage {
    /// Number of prompt tokens.
    pub prompt_tokens: u32,
    /// Total tokens.
    pub total_tokens: u32,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_embedding_request_creation() {
        let request = EmbeddingRequest::new("mistral-embed", "Hello, world!");
        assert_eq!(request.model, "mistral-embed");
        assert!(matches!(request.input, EmbeddingInput::Single(_)));
    }

    #[test]
    fn test_embedding_input_from_vec() {
        let input: EmbeddingInput = vec!["Hello", "World"].into();
        assert!(matches!(input, EmbeddingInput::Multiple(_)));
    }

    #[test]
    fn test_embedding_request_with_format() {
        let request = EmbeddingRequest::new("mistral-embed", "test")
            .with_encoding_format(EncodingFormat::Float);
        assert_eq!(request.encoding_format, Some(EncodingFormat::Float));
    }
}
